### PR TITLE
drivers: ipm: Add support for NXP i.MX rev2 MU

### PR DESCRIPTION
**Pull request by trowbridgec1**
With a number of the i.MX SoCs (e.g., the i.MX8M Plus), NXP has moved to supporting the
integrated Cortex-M cores with the MCUXpresso SDK (MCUX). As a result, certain Zephyr
drivers (such as the IPM driver) need to be updated to utlize this new MCUX-based SDK.

This change adds support for the MCUX Messaging Unit driver pulled in by this PR:
https://github.com/zephyrproject-rtos/hal_nxp/pull/130

Additionally, this change enables the new `IPM_IMX_REV2` config for the `mimx8ml8_m7` SoC
target which utilizes this new revision of the driver.

Signed-off-by: Chris Trowbridge <chris.trowbridge@lairdconnect.com>

### CODE REVIEW

**Comment by carlescufi** [comment]
> This change adds support for the MCUX Messaging Unit driver pulled in by this PR:
> [zephyrproject-rtos/hal_nxp#130](https://github.com/zephyrproject-rtos/hal_nxp/pull/130)

Did the manifest (`west.yml`) get updated to point to the new SHA in hal_nxp?

**Comment by trowbridgec-laird** [comment]
@carlescufi: yes - The revision for `hal_nxp` in `west.yml` was updated to `9629f15fd80169cc9669e3fbb1260467c3aeb175` in PR https://github.com/zephyrproject-rtos/zephyr/pull/42270, and taking a look at [`mcux/hal_nxp.cmake` at that revision](https://github.com/zephyrproject-rtos/hal_nxp/blob/9629f15fd80169cc9669e3fbb1260467c3aeb175/mcux/hal_nxp.cmake#L85), you can see that the support for this new config (`CONFIG_IPM_IMX_REV2`) is present:

```
...
include_driver_ifdef(CONFIG_ADC_MCUX_ADC12		adc12		driver_adc12)
include_driver_ifdef(CONFIG_ADC_MCUX_ADC16		adc16		driver_adc16)
include_driver_ifdef(CONFIG_IPM_IMX_REV2		mu		driver_mu)
include_driver_ifdef(CONFIG_CAN_MCUX_FLEXCAN		flexcan		driver_flexcan)
include_driver_ifdef(CONFIG_COUNTER_MCUX_PIT		pit		driver_pit)
...
```

**Comment by dornbirndevelops** [comment]
@trowbridgec-laird  these pending changes were the ones I was looking for the past two weeks (though I have a different board).
Thank you very much for that.

**Review by danieldegrasse — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

File: drivers/ipm/ipm_imx.c [hunk 1]
```diff
@@ -81,7 +148,7 @@ static void imx_mu_isr(const struct device *dev)
 	 * Store immediate overlapping exception return operation
 	 * might vector to incorrect interrupt
 	 */
-#if defined __CORTEX_M && (__CORTEX_M == 4U)
+#if (defined __CORTEX_M) && ((__CORTEX_M == 4U) || (__CORTEX_M == 7U))
```
Comment (emfend) [inline drivers/ipm/ipm_imx.c:151]: I was wondering why __DSB is also required for Cortex-M7.
While searching for possible reasons, I came across this comment:
```
    /* ARM errata 838869, affects Cortex-M4, Cortex-M4F Store immediate overlapping
     * exception return operation might vector to incorrect interrupt.
     * For Cortex-M7, if core speed much faster than peripheral register write speed,
     * the peripheral interrupt flags may be still set after exiting ISR, this results to
     * the same error similar with errata 83869 */
```
If that was the reason to enable this for M7 as well, it would be nice to mention that in the commit.

Thanks for your work!

Comment (trowbridgec-laird) [reply drivers/ipm/ipm_imx.c:151]: @emfend Good find! When adding this support for the rev2 MU peripheral, I used the RPMsg-Lite examples from the MCUXpresso SDK for this target (the i.MX8M Plus SoC) as a reference, and that comment is from the MU IRQ handler in [`rpmsg_platform.c`](https://github.com/NXPmicro/rpmsg-lite/blob/master/lib/rpmsg_lite/porting/platform/imx8mp_m7/rpmsg_platform.c#L89-L111):

```c++
/*
 * MU Interrrupt RPMsg handler
 */
int32_t MU1_M7_IRQHandler(void)
{
    uint32_t channel;

    if ((((1UL << 27U) >> RPMSG_MU_CHANNEL) & MU_GetStatusFlags(MUB)) != 0UL)
    {
        channel = MU_ReceiveMsgNonBlocking(MUB, RPMSG_MU_CHANNEL); // Read message from RX register.
        env_isr(channel >> 16);
    }
    /* ARM errata 838869, affects Cortex-M4, Cortex-M4F Store immediate overlapping
     * exception return operation might vector to incorrect interrupt.
     * For Cortex-M7, if core speed much faster than peripheral register write speed,
     * the peripheral interrupt flags may be still set after exiting ISR, this results to
     * the same error similar with errata 83869 */
#if (defined __CORTEX_M) && ((__CORTEX_M == 4U) || (__CORTEX_M == 7U))
    __DSB();
#endif

    return 0;
}
```

I'll add the second half the comment from `rpmsg_platform.c` to the one in this file like you suggest. Thanks!

Comment (emfend) [reply drivers/ipm/ipm_imx.c:151]: I came across this while writing a driver [1] for exactly the same MU :)

Since I was afraid that it would become confusing if another driver was packed into the existing file, I decided to pack this into a separate file. But from what I've seen, it's quite ok.
Since you were faster, your version will go upstream anyway.
Thanks for your work!

[1] https://github.com/emfend/zephyr/tree/imx8mp-mu

File: soc/arm/nxp_imx/mimx8ml8_m7/Kconfig.defconfig.mimx8ml8_m7 [hunk 2]
```diff
@@ -34,6 +34,10 @@ config UART_MCUX_IUART
 
 endif # SERIAL
 
+config IPM_IMX_REV2
+	default y
+	depends on IPM
+
```
Comment (pdgendt) [inline soc/arm/nxp_imx/mimx8ml8_m7/Kconfig.defconfig.mimx8ml8_m7:40]: Maybe move this to a separate commit, also shouldn't the `compatible = "nxp,imx-mu";` entries be added to `nxp_imx8ml_m7.dtsi` to add the actual devices?
